### PR TITLE
Add backpressure information to tee, clone

### PR DESCRIPTION
**Pull request by yonran**
Add information on what causes backpressure to ReadableStream.tee, Request.clone, Response.clone.

#### Summary

Add warning that `ReadableStream.tee`, `Request.clone`, `Response.clone` will backpressure to the speed of the faster consumer, and unread data will be buffered at the slower `ReadableStream` without limit and without backpressure.

#### Motivation

I researched this issue when I studied the discrepancy between nodejs implementations such as node-fetch and minipass-fetch based on `Readable.pipe` (which backpressures to the slower consumed stream to limit buffering) and browsers (which backpressure to the faster consumed stream and have unlimited buffering). I wrote my findings in https://github.com/node-fetch/node-fetch/issues/1568

#### Supporting details

The original sin was that the implementation of tee in the streams spec backpressured to the faster rather than the slower stream https://github.com/whatwg/streams/pull/311, which contradicted the bounded-buffering suggestion in [Stream requirements:. You must be able to pipe a stream to more than one writable stream.](https://github.com/whatwg/streams/blob/e9355ce79925947e8eb496563d599c329769d315/Requirements.md#you-must-be-able-to-pipe-a-stream-to-more-than-one-writable-stream) which said, “The tee stream can use a number of strategies to govern how the speed of its outputs affect the backpressure signals it gives off, but  the simplest strategy is to pass aggregate backpressure signals directly up the chain, thus letting the speed of the slowest output determine the speed of the tee.”

The informative text in the streams spec is ambiguous; it suggests that tee is useful for “independent” consumers ([Streams Specification: 2.1. Readable Streams](https://streams.spec.whatwg.org/#rs-model), [Streams Specification 4.2.4. ReadableStream example with tee](https://streams.spec.whatwg.org/#example-tee-and-pipe)). But `tee()` is dangerous to use for independent consumers, since the slower consumer buffers without backpressure or limit and will use up all your memory and crash. But if we can’t fix `tee()`, at least we should document its shortcomings.
.
#### Metadata

- [ ] Adds a new document
- [ ] Rewrites (or significantly expands) a document
- [ ] Fixes a typo, bug, or other error

### CODE REVIEW

File: files/en-us/web/api/response/clone/index.md [hunk 1]
```diff
@@ -14,6 +14,15 @@ browser-compat: api.Response.clone
 
 The **`clone()`** method of the {{domxref("Response")}} interface creates a clone of a response object, identical in every way, but stored in a different variable.
 
+Like the underlying {{domxref("ReadableStream.tee")}} api,
+the {{domxref("Response.body", "body")}} of a cloned `Response`
+will backpressure to the speed of the *faster* consumed `ReadableStream`,
+and unread data is buffered onto the internal buffer
+of the slower consumed `ReadableStream` without any limit or backpressure.
+If only one branch is consumed, then the entire body will be buffered in memory.
+Therefore, you should not use the build-in `clone()` to read very large bodies
```
Comment (OnkarRuikar) [inline files/en-us/web/api/response/clone/index.md:23]:
```suggestion
Therefore, you should not use the built-in `clone()` to read very large bodies
```

File: files/en-us/web/api/readablestream/tee/index.md [hunk 2]
```diff
@@ -17,12 +17,22 @@ The **`tee()`** method of the
...
-This is useful for allowing two readers to read a stream simultaneously, perhaps at
-different speeds. You might do this for example in a ServiceWorker if you want to fetch
+This is useful for allowing two readers to read a stream sequentially or simultaneously,
+perhaps at different speeds.
+For example, you might do this in a ServiceWorker if you want to fetch
 a response from the server and stream it to the browser, but also stream it to the
 ServiceWorker cache. Since a response body cannot be consumed more than once, you'd need
 two copies to do this.
 
+A teed stream will backpressure to the speed of the *faster* consumed `ReadableStream`,
```
Comment (teoli2003) [inline files/en-us/web/api/readablestream/tee/index.md:27]: Drive-by comment here.

Even if I know well how connections and messages do work (I was able to manually decode TCP messages back then), as a non-native reader, I find the term _backpressure_ difficult to understand. I think we should find a better term, or at least succintly define it the first time we use it in a page.

Comment (yonran) [reply files/en-us/web/api/readablestream/tee/index.md:27]: Fair point. I realize that I was using backpressure as a verb like in the akka-streams documentation e.g. [Source.alsoTo](https://doc.akka.io/docs/akka/current/stream/operators/Source-or-Flow/alsoTo.html). But in [nodejs stream](https://nodejs.org/api/stream.html) and [streams spec](https://streams.spec.whatwg.org/), backpressure is only used as a noun. Also the growable internal queue and support for both [pull sources](https://streams.spec.whatwg.org/#pull-source) and [push sources](https://streams.spec.whatwg.org/#push-source) when implementing an [underlying source](https://streams.spec.whatwg.org/#underlying-source-api) make backpressure more complicated.

In the streams spec, backpressure is specifically the condition that `controller.desiredSize <= 0`, which a [push](https://streams.spec.whatwg.org/#push-source) [underlying source](https://streams.spec.whatwg.org/#underlying-source-api) is expected to pause calling [`controller.enqueue(chunk)`](https://streams.spec.whatwg.org/#rs-default-controller-enqueue), and the controller will stop repeatedly calling [`pull(controller)`](https://streams.spec.whatwg.org/#dom-underlyingsource-pull) to the [underlying source](https://streams.spec.whatwg.org/#underlying-source-api) which might be a [pull source](https://streams.spec.whatwg.org/#pull-source).

Comment (yonran) [reply files/en-us/web/api/readablestream/tee/index.md]: @teoli2003 I have added a technical explanation of backpressure in `ReadableStream.tee`, and also I opened a PR to the streams spec so hopefully the spec will match my description https://github.com/whatwg/streams/pull/1234.

I also added an informal description of backpressure to `Request.clone`.

File: files/en-us/web/api/request/clone/index.md [hunk 3]
```diff
@@ -14,6 +14,13 @@ browser-compat: api.Request.clone
 
 The **`clone()`** method of the {{domxref("Request")}} interface creates a copy of the current `Request` object.
 
+Like the underlying {{domxref("ReadableStream.tee")}} api,
+the {{domxref("Request.body", "body")}} of a cloned `Response`
+will signal backpressure at the rate of the *faster* consumer of the two bodies,
```
Comment (teoli2003) [inline files/en-us/web/api/request/clone/index.md:19]:
```suggestion
will signal backpressure at the rate of the _faster_ consumer of the two bodies,
```

File: files/en-us/web/api/readablestream/tee/index.md [hunk 4]
```diff
@@ -17,12 +17,29 @@ The **`tee()`** method of the
...
+For example, you might do this in a ServiceWorker if you want to fetch
 a response from the server and stream it to the browser, but also stream it to the
 ServiceWorker cache. Since a response body cannot be consumed more than once, you'd need
 two copies to do this.
 
+A teed stream will partially signal backpressure at the rate of the *faster* consumer
+of the two `ReadableStream` branches,
+and unread data is enqueued internally on the slower consumed `ReadableStream`
+without any limit or backpressure.
+That is, when *both* branches have an unread element in their internal queue,
```
Comment (teoli2003) [inline files/en-us/web/api/readablestream/tee/index.md:31]:
```suggestion
That is, when _both_ branches have an unread element in their internal queue,
```

**Review by teoli2003 — CHANGES_REQUESTED** [review]
Two minor changes and I think we can merge this.

File: files/en-us/web/api/readablestream/tee/index.md (hunk 2 above)
Comment (hamishwillee) [reply files/en-us/web/api/readablestream/tee/index.md]: This is interesting. I hadn't got to my reading around Tees yet!

FYI Just to be a little bit pedantic because my head is in this at the moment for byte streams work ...

A push underlying source does use backpressure to change how it enqueues new data - it will always enqueue whatever data it gets. If it gets "backpressure" it is supposed to signal the thing that is providing the data (e.g. a socket) that it should pause or throttle supply. The difference is that not every supply has a mechanism for throttling or pausing.

As you say, for a pull source the controller manages the pull requests based on need (spec appears to say that it keeps requesting until bufffers are filled (desired size = 0) then stops calling it until empty and more data is requested. Though it looks like there are different strategies you can use there.

Comment (hamishwillee) [reply files/en-us/web/api/readablestream/tee/index.md]: PS Backpressure is actually  not a bad term IMO, but as you say @teoli2003 it needs to be properly defined. Its a bit like a pipe - you can fill it up so much before the water starts pushing back up the pipe, and eventually you can't fill it any more.

File: files/en-us/web/api/readablestream/tee/index.md (hunk 4 above)
Comment (yonran) [reply files/en-us/web/api/readablestream/tee/index.md]: Sure. I converted `*x*` to `_x_` here and a couple other places

**Review by teoli2003 — APPROVED** [review]
👍

**Comment by yonran** [comment]
Do you want me to rebase and squash and add the PR number (#16084) to the commit message?

**Comment by teoli2003** [comment]
It's fine like this!

Thanks a lot!

Congratulations upon your first merged commit here! Welcome aboard! :tada:
